src/main/java/at/hannibal2/skyhanni/features/itemabilities/CrownOfAvariceCounter.kt

```diff
@@ -25,28 +35,117 @@ object CrownOfAvariceCounter {
     private val internalName = "CROWN_OF_AVARICE".toInternalName()
 
     private var render: Renderable? = null
+    private const val MAX_AVARICE = 1_000_000_000L
+    private val MAX_AFK_TIME = 2.minutes
+
+    private val isWearingCrown by RecalculatingValue(1.seconds) {
+        InventoryUtils.getHelmet()?.getInternalNameOrNull() == internalName
+    }
+
+    private var count: Long = 0L
+    private var coinsEarned: Long = 0L
+    private var sessionStart = SimpleTimeMark.farPast()
+    private var lastCoinUpdate = SimpleTimeMark.farPast()
+    private val isSessionActive get(): Boolean = sessionStart.passedSince() < 10.seconds
+    private var coinsDifference: Long = 0L
 
     @HandleEvent
     fun onRenderOverlay(event: GuiRenderEvent.GuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        if (!isWearingCrown) return
         render?.let { config.position.renderRenderable(it, posLabel = "Crown of Avarice Counter") }
     }
 
-    @SubscribeEvent
-    fun onTick(event: LorenzTickEvent) {
-        render = check()
+    @HandleEvent
+    fun onSecondPassed(event: SecondPassedEvent) {
+        if (!isEnabled()) return
+        if (!isWearingCrown) return
+        update()
+    }
+
+    @HandleEvent
+    fun onInventoryUpdated(event: OwnInventoryItemUpdateEvent) {
+        if (!isEnabled() || event.slot != 5) return
+        val item = event.itemStack
+        if (item.getInternalNameOrNull() != internalName) return
+        val coins = item.getCoinsOfAvarice() ?: return
+        if (count == 0L) count = coins
+
+        coinsDifference = coins - count
+
+        if (coinsDifference == 0L) return
+
+        if (coinsDifference < 0) {
+            reset()
+            count = coins
+            return
+        }
+
+        if (isSessionAFK()) reset()
+        lastCoinUpdate = SimpleTimeMark.now()
+        coinsEarned += coinsDifference
+        count = coins
+
+        update()
     }
 
-    fun check(): Renderable? {
-        if (!LorenzUtils.inSkyBlock) return null
-        if (!config.enable) return null
-        val item = InventoryUtils.getHelmet()
-        if (item?.getInternalNameOrNull() != internalName) return null
-        val count = item.extraAttributes.getLong("collected_coins")
-        return Renderable.horizontalContainer(
-            listOf(
-                Renderable.itemStack(internalName.getItemStack()),
-                Renderable.string("§6" + if (config.shortFormat) count.shortFormat() else count.addSeparators()),
-            ),
-        )
+    @HandleEvent
+    fun onIslandChange(event: IslandChangeEvent) {
+        reset()
+        count = InventoryUtils.getHelmet()?.getCoinsOfAvarice() ?: return
+    }
+
+    private fun update() {
+        render = Renderable.verticalContainer(buildList())
```

Comment:
is there a reason why we use a vertical container? if no, please remove the wrapper, and render a list of renderable instead (instead of null, you work with empty)

Reply:
Also please use consistent variable names, the thing that gets rendered is called `display` all over the mod